### ISSUE TITLE
Unable to find "/v1, Resource=pods" invalid label key "-ltier"

### ISSUE THREAD

**Issue by rally25rs** [groundnuty/k8s-wait-for#19]
I'm getting this error:

```
No resources found.
FalseError from server (BadRequest): Unable to find "/v1, Resource=pods" that match label selector "app=cloud-services,-ltier=postgres", field selector "": unable to parse requirement: invalid label key "-ltier": name part must consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyName',  or 'my.name',  or '123-abc', regex used for validation is '([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]')
```

This is on container image tag `v1.2-5-g92c083e` and the current script in `master`.

-----

This appears to be caused by
```
    get_service_state_selectors=$(kubectl get service "$get_service_state_name" $KUBECTL_ARGS -ojson 2>&1 | jq -cr 'if . | has("items") then .items[]
else . end | [ .spec.selector | to_entries[] | "-l\(.key)=\(.value)" ] | join(",") ')
```
when the service has more than 1 selector

```
        "selector": {
            "app": "cloud-services",
            "tier": "postgres"
```

The `to_entries[] | "-l\(.key)=\(.value)" ] | join(",")` appears to be incorrect and would make

```
-lapp=cloud-services,-ltier=postgres
```
when it should not have the additional `-l`. It should be
```
-lapp=cloud-services,tier=postgres
```

The code below that is:
```
    for get_service_state_selector in $get_service_state_selectors ; do
        get_service_state_selector=$(echo "$get_service_state_selector" | tr ',' ' ')
        get_service_state_state=$(get_pod_state "$get_service_state_selectors")
        get_service_state_states="${get_service_state_states}${get_service_state_state}" ;
    done
    echo "$get_service_state_states"
```
which looks like it would then replace that comma with a space to make
```
-lapp=cloud-services -ltier=postgres
```
except for 2 issues; 1) `get_service_state_selector` is never used, only assigned to, 2) I don't think the kubectl command will allow multiple `-l` arguments anyway.

-----

I _think_ this could be fixed by doing:

```
get_service_state() {
    get_service_state_name="$1"
    get_service_state_selectors=$(kubectl get service "$get_service_state_name" $KUBECTL_ARGS -ojson 2>&1 | jq -cr 'if . | has("items") then .items[]
else . end | [ .spec.selector | to_entries[] | "\(.key)=\(.value)" ] | join(",") ')
    get_service_state_states=""
    for get_service_state_selector in $get_service_state_selectors ; do
        get_service_state_state=$(get_pod_state "-l$get_service_state_selectors")
        get_service_state_states="${get_service_state_states}${get_service_state_state}" ;
    done
    echo "$get_service_state_states"
}
```

But I'm not sure of all the cases that this has to handle.

**Comment by Dlougach** [comment]
This bug isn't fixed in groundnuty repo.

**Comment by rally25rs** [comment]
yeah it looks like GitHub marked this issue closed when I merged the fix into my fork.
I'll re-open this issue.
My fix for it is in #20 which has not been merged.

---

Bit of a shameless plug, but I didn't even use this library after making this fix, because it doesn't really work well with a cluster that uses RBAC. I wrote another tool that accommodated my needs better: https://github.com/rally25rs/k8s-when-ready
